src/server.c

```diff
@@ -3176,36 +3203,48 @@ void updateCommandLatencyHistogram(struct hdr_histogram **latency_histogram, int
  * multiple separated commands. Note that alsoPropagate() is not affected
  * by CLIENT_PREVENT_PROP flag. */
 void propagatePendingCommands() {
-    if (server.also_propagate.numops == 0)
+    if (server.also_propagate.numops == 0) {
+        redisOpArrayFree(&server.also_propagate);
         return;
+    }
 
     int j;
     redisOp *rop;
+    redisOp *last_command = NULL;
+    redisOp *first_command = NULL;
     int multi_emitted = 0;
 
-    /* Wrap the commands in server.also_propagate array,
-     * but don't wrap it if we are already in MULTI context,
-     * in case the nested MULTI/EXEC.
-     *
-     * And if the array contains only one command, no need to
-     * wrap it, since the single command is atomic. */
-    if (server.also_propagate.numops > 1 && !server.propagate_no_multi) {
-        /* We use the first command-to-propagate to set the dbid for MULTI,
-         * so that the SELECT will be propagated beforehand */
-        int multi_dbid = server.also_propagate.ops[0].dbid;
-        propagateNow(multi_dbid,&shared.multi,1,PROPAGATE_AOF|PROPAGATE_REPL);
-        multi_emitted = 1;
-    }
-
-    for (j = 0; j < server.also_propagate.numops; j++) {
+    for (j = 0; j < server.also_propagate.len; j++) {
         rop = &server.also_propagate.ops[j];
-        serverAssert(rop->target);
+        if (!rop->target) {
+            continue;
+        }
+        if (!first_command) {
+            first_command = rop;
+            /* Wrap the commands in server.also_propagate array,
+             * but don't wrap it if we are already in MULTI context,
+             * in case the nested MULTI/EXEC.
+             *
+             * And if the array contains only one command, no need to
+             * wrap it, since the single command is atomic. */
+            if (server.also_propagate.numops > 1 && !server.propagate_no_multi) {
+                /* We use the first command-to-propagate to set the dbid for MULTI,
+                 * so that the SELECT will be propagated beforehand */
+                int multi_dbid = first_command->dbid;
+                propagateNow(multi_dbid,&shared.multi,1,PROPAGATE_AOF|PROPAGATE_REPL);
+                multi_emitted = 1;
+            }
+
+        }
+        last_command = rop;
         propagateNow(rop->dbid,rop->argv,rop->argc,rop->target);
     }
 
+    serverAssert(last_command);
+
     if (multi_emitted) {
         /* We take the dbid from last command so that propagateNow() won't inject another SELECT */
-        int exec_dbid = server.also_propagate.ops[server.also_propagate.numops-1].dbid;
+        int exec_dbid = last_command->dbid;
         propagateNow(exec_dbid,&shared.exec,1,PROPAGATE_AOF|PROPAGATE_REPL);
```

Comment:
please make propagateNow take dbid=-1, meaning it should not emit any SELECT (db-neutral command)